lib/OnyxUtils.ts

```diff
@@ -783,71 +711,59 @@ function keyChanged(key, data, prevData, canUpdateSubscriber = () => true, notif
  * Sends the data obtained from the keys to the connection. It either:
  *     - sets state on the withOnyxInstances
  *     - triggers the callback function
- *
- * @private
- * @param {Object} mapping
- * @param {Object} [mapping.withOnyxInstance]
- * @param {String} [mapping.statePropertyName]
- * @param {Function} [mapping.callback]
- * @param {String} [mapping.selector]
- * @param {*|null} val
- * @param {String|undefined} matchedKey
- * @param {Boolean} isBatched
  */
-function sendDataToConnection(mapping, val, matchedKey, isBatched) {
+function sendDataToConnection<TKey extends OnyxKey>(
+    mapping: Mapping<TKey>,
+    val: OnyxCollection<OnyxValue<TKey>> | OnyxEntry<OnyxValue<TKey>>,
+    matchedKey: TKey | undefined,
+    isBatched: boolean,
+): void {
     // If the mapping no longer exists then we should not send any data.
     // This means our subscriber disconnected or withOnyx wrapped component unmounted.
     if (!callbackToStateMapping[mapping.connectionID]) {
         return;
     }
 
-    if (mapping.withOnyxInstance) {
-        let newData = val;
+    if ('withOnyxInstance' in mapping) {
+        let newData: OnyxValue<OnyxKey> = val;
 
         // If the mapping has a selector, then the component's state must only be updated with the data
         // returned by the selector.
         if (mapping.selector) {
             if (isCollectionKey(mapping.key)) {
-                newData = reduceCollectionWithSelector(val, mapping.selector, mapping.withOnyxInstance.state);
+                newData = reduceCollectionWithSelector(val as OnyxCollection<OnyxValue<TKey>>, mapping.selector, mapping.withOnyxInstance.state);
```

Comment:
```suggestion
                newData = reduceCollectionWithSelector(val as OnyxCollection<KeyValueMapping[TKey]>, mapping.selector, mapping.withOnyxInstance.state);
```

but we need to fix here and `reduceCollectionWithSelector` to accept `OnyxCollection<KeyValueMapping[TKey]>`